### ISSUE TITLE
Go regression creates jitter in the timer, creating less compressible timestamps, which grows TSDB data size across the board

### ISSUE THREAD

**Issue by Seitanas** [prometheus/prometheus#7846]
Hello, 
we are having discussion on strange issue on data usage growth since upgrading Prometheus on victoria-metrics github page.
As it appeared at the beginning that it may be victoria-metrics problem, we saw that the downgrade to Prometheus 2.16.0 solves the issue. Perhaps you any ideas what is happening there?
Link to the discussin:
﻿https://github.com/VictoriaMetrics/VictoriaMetrics/issues/695

Thank you.

**Comment by roidelapluie** [comment]
We need more details about this setup, the versions, involved, the configs, to try to asses whether it is a regression of prometheus or not.

Also it seems that you have downgraded just 1 hour ago which might not be enough to show that it is enough.

**Comment by Seitanas** [comment]
Waited for almost 24hours, still looks that storage usage has improved:
<img width="516" alt="Screenshot 2020-08-26 at 10 37 04" src="https://user-images.githubusercontent.com/6713389/91274916-1b6e3f80-e788-11ea-9ddf-dc26a7d08bf7.png">

The setup:
we are using two prometheus machines that are scraping and sending same data (for ha purposes) every 30s to 3 independent instances of victoria metrics.
All scrape targets are seleced via consul.
We began seeing increase on used storage after upgrading from v2.16.0 to v2.19.2. We tried to upgrade to v2.20.1 after but got same results. After downgrading to v2.16.0 storage consumption slowed down.
Version of victoria-metrics was v1.35.6, then we tried to upgrade to v1.39.4 - also had no visible improvement.

**Comment by roidelapluie** [comment]
Like that I don't really have an idea, do you have node exporter reporting an increase of network traffic as well?

**Comment by roidelapluie** [comment]
Could you show us a printscreen of https://grafana.com/grafana/dashboards/12054 ?

**Comment by Seitanas** [comment]
<img width="1591" alt="Screenshot 2020-08-31 at 09 04 06" src="https://user-images.githubusercontent.com/6713389/91687669-f9e6cc80-eb68-11ea-9161-85bf8330395f.png">
And no - there was no traffic increase nor increase on datapoint count.

More interesting I've noticed - prometheus local sore also began to grow after upgrade. And when I've downgraded to 2.16.0  usage slowly began to drop (due retention, which is 5days):

<img width="1591" alt="Screenshot 2020-08-31 at 09 06 37" src="https://user-images.githubusercontent.com/6713389/91687789-4cc08400-eb69-11ea-9439-f88082bf7b4d.png">

**Comment by gouthamve** [comment]
Hrm looking at this, this could be due to us mmap-ing headchunks to disk. But this would explain the slight increase in disk usage in Prometheus, but this should not affect the remote-write at all. Could you see if the data in VM is compressing just as well as it did when running with 2.16? Could you also check if the number of samples being sent is the same in both cases?

This looks like an issue in VictoriaMetrics more than Prometheus, but we'd appreciate more info if you have any.

**Comment by Seitanas** [comment]
I can see no visible effect on ingestion rate:
<img width="786" alt="Screenshot 2020-08-31 at 17 46 22" src="https://user-images.githubusercontent.com/6713389/91733205-ed3a9680-ebb1-11ea-95e5-f345d3e77c5b.png">
Nor in merge part:
<img width="1585" alt="Screenshot 2020-08-31 at 17 47 57" src="https://user-images.githubusercontent.com/6713389/91733404-2410ac80-ebb2-11ea-827e-847c937f3e64.png">
Can't find any metrics to represent compression though.
The only thing that did change after upgrade was `bytes-per-datapoint` size began to drop (and began to rise again after downgrade):

<img width="875" alt="Screenshot 2020-08-31 at 17 49 06" src="https://user-images.githubusercontent.com/6713389/91733577-5f12e000-ebb2-11ea-8c84-c75884c66cd6.png">

**Comment by Seitanas** [comment]
We made a test environment consisting of 2 VM servers and 2 Prometheus servers to find out on which prometheus version this issue began. So it seems that all Prometheus versions prio to 2.18.0 are using normal amounts of space. We also noticed that not only VM storage grows, - prometheus storage grows also:

**Prometheus 2.16.0 VS prometheus 2.17.2:**

VM data usage:
<img width="1591" alt="2 16 0-2 17 2" src="https://user-images.githubusercontent.com/6713389/93064635-6ccf6780-f680-11ea-9c44-a31b0015baa7.png">

Prometheus data usage:
<img width="1591" alt="2 16 0 -2 17 2-1" src="https://user-images.githubusercontent.com/6713389/93064653-7062ee80-f680-11ea-8f8c-25dcec1b6c2b.png">

**Prometheus 2.16.0 VS prometheus 2.18.0:**

VM data usage:
<img width="1591" alt="2 16 0-2 18 0" src="https://user-images.githubusercontent.com/6713389/93064562-52958980-f680-11ea-9883-a0cddaffca78.png">

Prometheus data usage:
<img width="1591" alt="2 16 0-2 18 0-1" src="https://user-images.githubusercontent.com/6713389/93064571-55907a00-f680-11ea-8f54-ba06cde12273.png">

**Comment by roidelapluie** [comment]
Could you try running prometheus with --no-storage.tsdb.wal-compression and see if that makes a difference?

**Comment by roidelapluie** [comment]
oh 2.18 ..
thanks, in 2.18 there is no such comprssion by default

**Comment by roidelapluie** [comment]
Possible duplicate of #7973

**Comment by roidelapluie** [comment]
Hello,

This might be a bug in go, causing scrapes not to be aligned correctly:
https://github.com/golang/go/issues/38860

**Comment by roidelapluie** [comment]
```
rate﻿(﻿prometheus_tsdb_compaction_chunk_size_bytes_sum﻿{﻿job﻿=﻿"prometheus"﻿}﻿[﻿24h﻿]﻿)﻿ / ﻿rate﻿(﻿prometheus_tsdb_compaction_chunk_samples_sum﻿{﻿job﻿=﻿"prometheus"﻿}﻿[﻿24h﻿]﻿)
```

should show a difference:

![compression](https://user-images.githubusercontent.com/291750/94490956-8ecffa80-01e7-11eb-9c58-559f5bb05607.png)

**Comment by roidelapluie** [comment]
The small improvement in 2.21 can be explained by https://go-review.googlesource.com/c/go/+/216198/

**Comment by Seitanas** [comment]
Hello,
I've  ran tests against https://github.com/prometheus/prometheus/pull/7976
and I can confirm, that the fix solves https://github.com/VictoriaMetrics/VictoriaMetrics/issues/695 problem:

Red line Prometheuss `2.21.0 / 2020-09-11`
Green line - `PR7976` compiled with `Go 1.15`
<img width="1614" alt="Screenshot 2020-09-29 at 07 38 24" src="https://user-images.githubusercontent.com/6713389/94513501-88fa0980-0227-11eb-91a0-b6b5b7299100.png">

**Comment by roidelapluie** [comment]
#7981 might help too, but I have still to test it.

**Comment by roidelapluie** [comment]
Encouraging results in prod too

![blocks](https://user-images.githubusercontent.com/291750/95441770-d4fd2a80-095a-11eb-8fa5-320b9549a2e5.png)

**Comment by roidelapluie** [comment]
Closed by #7973

Followup discussions on the bugfix itself #8032
